File: csharp/src/LibUSearch/USearchIndex.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using static NativeMethods;

public class USearchIndex : IDisposable
{
    private IntPtr _index;
    private bool _disposedValue = false;
    private ulong _cached_dimensions;

    public USearchIndex(
        usearch_metric_kind_t metricKind,
        usearch_scalar_kind_t quantization,
        ulong dimensions,
        ulong connectivity = 0,
        ulong expansionAdd = 0,
        ulong expansionSearch = 0
    // CustomDistanceFunction? customMetric = null
    )
    {
        usearch_init_options_t initOptions = new usearch_init_options_t
        {
            metric_kind = metricKind,
            metric = default,
            quantization = quantization,
            dimensions = (UIntPtr)dimensions,
            connectivity = (UIntPtr)connectivity,
            expansion_add = (UIntPtr)expansionAdd,
            expansion_search = (UIntPtr)expansionSearch
        };

        this._index = usearch_init(ref initOptions, out IntPtr error);
        HandleError(error);
        this._cached_dimensions = dimensions;
    }

    public USearchIndex(usearch_init_options_t options)
    {
        usearch_init_options_t initOptions = options;
        this._index = usearch_init(ref initOptions, out IntPtr error);
        HandleError(error);
        this._cached_dimensions = (ulong)options.dimensions;
    }

    public USearchIndex(string path, bool view = false)
    {
        usearch_init_options_t initOptions = new usearch_init_options_t();
        this._index = usearch_init(ref initOptions, out IntPtr error);
        HandleError(error);

        if (view)
        {
            usearch_view(this._index, path, out error);
        }
        else
        {
            usearch_load(this._index, path, out error);
        }

        HandleError(error);

        this._cached_dimensions = this.Dimensions();
    }

    public void Save(string path)
    {
        usearch_save(this._index, path, out IntPtr error);
        HandleError(error);
    }

    public ulong Size()
    {
        ulong size = (ulong)usearch_size(this._index, out IntPtr error);
        HandleError(error);
        return size;
    }

    public ulong Capacity()
    {
        ulong capacity = (ulong)usearch_capacity(this._index, out IntPtr error);
        HandleError(error);
        return capacity;
    }

    public ulong Dimensions()
    {
        ulong dimensions = (ulong)usearch_dimensions(this._index, out IntPtr error);
        HandleError(error);
        return dimensions;
    }

    public ulong Connectivity()
    {
        ulong connectivity = (ulong)usearch_connectivity(this._index, out IntPtr error);
        HandleError(error);
        return connectivity;
    }

    public bool Contains(ulong key)
    {
        bool result = usearch_contains(this._index, key, out IntPtr error);
        HandleError(error);
        return result;
    }

    private void IncreaseCapacity(ulong size)
    {
        usearch_reserve(this._index, (UIntPtr)(this.Size() + size), out IntPtr error);
        HandleError(error);
    }

    private void CheckIncreaseCapacity(ulong size_increase)
    {
        ulong size_demand = this.Size() + size_increase;
        if (this.Capacity() < size_demand)
        {
            this.IncreaseCapacity(size_increase);
        }
    }

    public void Add(ulong key, float[] vector)
    {
        this.CheckIncreaseCapacity(1);
        usearch_add(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f32_k, out IntPtr error);
        HandleError(error);
    }

    public void Add(ulong key, double[] vector)
    {
        this.CheckIncreaseCapacity(1);
        usearch_add(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f64_k, out IntPtr error);
        HandleError(error);
    }

    public void AddMany(ulong[] keys, float[][] vectors)
    {
        this.CheckIncreaseCapacity((ulong)vectors.Length);
        for (int i = 0; i < vectors.Length; i++)
        {
            usearch_add(this._index, keys[i], vectors[i], usearch_scalar_kind_t.usearch_scalar_f32_k, out IntPtr error);
            HandleError(error);
        }
    }

    public void AddMany(ulong[] keys, double[][] vectors)
    {
        this.CheckIncreaseCapacity((ulong)vectors.Length);
        for (int i = 0; i < vectors.Length; i++)
        {
            usearch_add(this._index, keys[i], vectors[i], usearch_scalar_kind_t.usearch_scalar_f64_k, out IntPtr error);
            HandleError(error);
        }
    }

    public bool Get(ulong key, out float[] vector)
    {
        vector = new float[this._cached_dimensions];
        bool success = usearch_get(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f32_k, out IntPtr error);
        HandleError(error);
        if (!success)
        {
            vector = null;
        }

        return success;
    }

    public bool Get(ulong key, out double[] vector)
    {
        vector = new double[this._cached_dimensions];
        bool success = usearch_get(this._index, key, vector, usearch_scalar_kind_t.usearch_scalar_f64_k, out IntPtr error);
        HandleError(error);
        if (!success)
        {
            vector = null;
        }

        return success;
    }

    private ulong Search<T>(T[] queryVector, ulong resultsLimit, out Dictionary<ulong, float> foundKeyDistances, usearch_scalar_kind_t scalarKind)
    {
        ulong[] keys = new ulong[resultsLimit];
        float[] distances = new float[resultsLimit];

        GCHandle handle = GCHandle.Alloc(queryVector, GCHandleType.Pinned);
        ulong matches = 0;
        try
        {
            IntPtr queryVectorPtr = handle.AddrOfPinnedObject();
            matches = (ulong)usearch_search(this._index, queryVectorPtr, scalarKind, (UIntPtr)resultsLimit, keys, distances, out IntPtr error);
            HandleError(error);
        }
        finally
        {
            handle.Free();
        }
        foundKeyDistances = new Dictionary<ulong, float>();
        for (ulong i = 0; i < matches; i++)
        {
            foundKeyDistances.Add(keys[i], distances[i]);
        }

        return matches;
    }

    public ulong Search(float[] queryVector, ulong resultsLimit, out Dictionary<ulong, float> foundKeyDistances)
    {
        return this.Search(queryVector, resultsLimit, out foundKeyDistances, usearch_scalar_kind_t.usearch_scalar_f32_k);
    }

    public ulong Search(double[] queryVector, ulong resultsLimit, out Dictionary<ulong, float> foundKeyDistances)
    {
        return this.Search(queryVector, resultsLimit, out foundKeyDistances, usearch_scalar_kind_t.usearch_scalar_f64_k);
    }

    public bool Remove(ulong key)
    {
        bool success = usearch_remove(this._index, key, out IntPtr error);
        HandleError(error);
        return success;
    }

    private static void HandleError(IntPtr error)
    {
        if (error != IntPtr.Zero)
        {
            Console.WriteLine($"Error {error}");
            throw new USearchException($"USearch operation failed: {Marshal.PtrToStringAnsi(error)}");
        }
    }

    private void FreeIndex()
    {
        if (this._index != IntPtr.Zero)
        {
            usearch_free(this._index, out IntPtr error);
            HandleError(error);
            this._index = IntPtr.Zero;
        }
    }

    public void Dispose()
    {
        this.Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (!this._disposedValue)
        {
            this.FreeIndex();
            this._disposedValue = true;
        }
    }

    ~USearchIndex() => this.Dispose(false);

    public class USearchException : Exception { public USearchException(string message) : base(message) { } }
}

File: swift/USearchIndex.swift
//
//  USearchIndex.swift
//
//
//  Created by Dan Palmer on 9/2/25.
//

import Foundation
import usearch_c

public enum USearchScalar: UInt {
    case f32
    case f16
    case f64
    case i8
    case b1
}

public enum USearchMetric: UInt {
    case ip
    case cos
    case l2sq
    case pearson
    case haversine
    case divergence
    case jaccard
    case hamming
    case tanimoto
    case sorensen
}

public typealias USearchKey = UInt64
public typealias USearchFilterFn = (USearchKey) -> Bool

extension USearchMetric {
    func toNative() -> usearch_metric_kind_t {
        switch self {
        case .ip:
            return usearch_metric_ip_k
        case .cos:
            return usearch_metric_cos_k
        case .l2sq:
            return usearch_metric_l2sq_k
        case .hamming:
            return usearch_metric_hamming_k
        case .haversine:
            return usearch_metric_haversine_k
        case .divergence:
            return usearch_metric_divergence_k
        case .jaccard:
            return usearch_metric_jaccard_k
        case .pearson:
            return usearch_metric_pearson_k
        case .sorensen:
            return usearch_metric_sorensen_k
        case .tanimoto:
            return usearch_metric_tanimoto_k
        }
    }
}

extension USearchScalar {
    func toNative() -> usearch_scalar_kind_t {
        switch self {
        case .i8:
            return usearch_scalar_i8_k
        case .f16:
            return usearch_scalar_f16_k
        case .b1:
            return usearch_scalar_b1_k
        case .f32:
            return usearch_scalar_f32_k
        case .f64:
            return usearch_scalar_f64_k
        }
    }
}

public enum USearchError: Error {
    case outOfMemory
    case unknownScalarKind
    case unknownMetricKind
    case fileReadError
    case fileTypeError
    case keyLookupsDisabled
    case keyMissing
    case serializationError
    case deserializationError
    case immutableError
    case keyCountError
    case renameCollisionError
    case indexTooSmallToClusterError
    case duplicateKeysError
    case reservationError

    case pathNotUTF8Encodable

    case unknownError(String)

    static func fromErrorString(_ errorString: String) -> USearchError {
        switch errorString {
        case
            "Out of memory!",
            "Out of memory",
            "Out of memory when preparing contexts!",
            "Out of memory, allocating a temporary buffer for batch results",
            "Failed to allocate memory for the index!",
            "Failed to allocate memory for the casts",
            "Failed to reserve memory for the index",
            "Failed to allocate memory for the available threads!",
            "Failed to allocate memory for the index",
            "Can't allocate memory for a free-list",
            "Failed to allocate memory for the casts!":
            return .outOfMemory
        case "Unknown scalar kind!":
            return .unknownScalarKind
        case "Unknown metric kind!":
            return .unknownMetricKind
        case "End of file reached!", "Can't infer file size":
            return .fileReadError
        case "Not a dense USearch index!":
            return .fileTypeError
        case "Key lookups are disabled!":
            return .keyLookupsDisabled
        case "Key missing!":
            return .keyMissing
        case "Failed to serialize into stream":
            return .serializationError
        case
            "Failed to read 32-bit dimensions of the matrix",
            "Failed to read 64-bit dimensions of the matrix",
            "Failed to allocate memory to address vectors",
            "Failed to read vectors",
            "Failed to read the index ", // space left intentionally blank
            "Magic header mismatch - the file isn't an index",
            "File format may be different, please rebuild",
            "Key type doesn't match, consider rebuilding",
            "Slot type doesn't match, consider rebuilding",
            "Index size and the number of vectors doesn't match",
            "File is corrupted and lacks matrix dimensions",
            "File is corrupted and lacks a header":
            return .deserializationError
        case
            "Can't add to an immutable index",
            "Can't remove from an immutable index":
            return .immutableError
        case "Free keys count mismatch":
            return .keyCountError
        case "Renaming impossible, the key is already in use":
            return .renameCollisionError
        case "Index too small to cluster!":
            return .indexTooSmallToClusterError
        case "Duplicate keys not allowed in high-level wrappers":
            return .duplicateKeysError
        case "Reserve capacity ahead of insertions!":
            return .reservationError
        default:
            return .unknownError(errorString)
        }
    }
}


@available(iOS 13.0, macOS 11.0, tvOS 13.0, watchOS 6.0, *)
public class USearchIndex: NSObject {
    private var nativeIndex: usearch_index_t

    private init(native: consuming usearch_index_t) throws {
        nativeIndex = native
        super.init()
    }

    var dimensions: UInt32 {
        get throws {
            return try UInt32(throwing { usearch_dimensions(nativeIndex, $0) })
        }
    }

    var connectivity: UInt32 {
        get throws {
            return try UInt32(throwing { usearch_connectivity(nativeIndex, $0) })
        }
    }

    var expansionAdd: UInt32 {
        get throws {
            return try UInt32(throwing { usearch_expansion_add(nativeIndex, $0) })
        }
    }

    var expansionSearch: UInt32 {
        get throws {
            return try UInt32(throwing { usearch_expansion_search(nativeIndex, $0) })
        }
    }

    var length: UInt32 {
        get throws {
            return try UInt32(throwing { usearch_size(nativeIndex, $0) })
        }
    }

    var capacity: UInt32 {
        get throws {
            return try UInt32(throwing { usearch_capacity(nativeIndex, $0) })
        }
    }

    /**
     * @brief Initializes a new index.
     * @param metric The distance function to compare the dis-similarity of vectors.
     * @param dimensions The number of dimensions planned for this index.
     * @param connectivity Number of connections per node in the proximity graph.
     * Higher connectivity improves quantization, increases memory usage, and reduces construction speed.
     * @param quantization Quantization of internal vector representations. Lower quantization means higher speed.
     */
    public static func make(metric: USearchMetric, dimensions: UInt32, connectivity: UInt32, quantization: USearchScalar) throws -> USearchIndex {
        return try make(metric: metric, dimensions: dimensions, connectivity: connectivity, quantization: quantization, multi: false)
    }

    /**
     * @brief Initializes a new index.
     * @param metric The distance function to compare the dis-similarity of vectors.
     * @param dimensions The number of dimensions planned for this index.
     * @param connectivity Number of connections per node in the proximity graph.
     * Higher connectivity improves quantization, increases memory usage, and reduces construction speed.
     * @param quantization Quantization of internal vector representations. Lower quantization means higher speed.
     * @param multi Enables indexing multiple vectors per key when true.
     */
    public static func make(metric: USearchMetric, dimensions: UInt32, connectivity: UInt32, quantization: USearchScalar, multi: Bool) throws -> USearchIndex {
        let options = usearch_init_options_t(
            metric_kind: metric.toNative(),
            metric: nil,
            quantization: quantization.toNative(),
            dimensions: Int(dimensions),
            connectivity: Int(connectivity),
            expansion_add: 0,
            expansion_search: 0,
            multi: multi
        )


        let optionsPtr = pointer(options)
        let index = try throwing { usearch_init(optionsPtr, $0) }
        guard let index else {
            throw USearchError.unknownError("No index created, no error returned.")
        }

        return try USearchIndex(native: index)
    }

    public var isEmpty: Bool {
        get throws {
            return (try throwing { usearch_size(nativeIndex, $0) } != 0)
        }
    }

    override public var description: String {
        do {
            return try "USearchIndex(dimensions: \(dimensions), connectivity: \(connectivity), length: \(length), capacity: \(capacity), isEmpty: \(isEmpty))"
        } catch {
            return "USearchIndex(error: \(error))"
        }
    }

    /**
     * @brief Pre-allocates space in the index for the given number of vectors.
     */
    public func reserve(_ count: UInt32) throws {
        try throwing { usearch_reserve(nativeIndex, Int(count), $0) }
    }

    /**
     * @brief Adds a labeled vector to the index.
     * @param vector Single-precision vector.
     */
    public func addSingle(key: USearchKey, vector: UnsafePointer<Float32>) throws {
        try throwing { usearch_add(nativeIndex, key, vector, USearchScalar.f32.toNative(), $0) }
    }

    /**
     * @brief Approximate nearest neighbors search.
     * @param vector Single-precision query vector.
     * @param count Upper limit on the number of matches to retrieve.
     * @param keys Optional output buffer for keys of approximate neighbors.
     * @param distances Optional output buffer for (increasing) distances to approximate neighbors.
     * @return Number of matches exported to `keys` and `distances`.
     */
    public func searchSingle(vector: UnsafePointer<Float32>, count: UInt32, keys: UnsafeMutablePointer<USearchKey>?, distances: UnsafeMutablePointer<Float32>?) throws -> UInt32 {
        let found = try throwing { usearch_search(nativeIndex, vector, USearchScalar.f32.toNative(), Int(count), keys, distances, $0) }
        return UInt32(found)
    }

    /**
    * @brief Retrieves a labeled single-precision vector from the index.
    * @param vector A buffer to store the vector.
    * @param count For multi-indexes, the number of vectors to retrieve.
    * @return Number of vectors exported to `vector`.
    */
    public func getSingle(key: USearchKey, vector: UnsafeMutablePointer<Float32>, count: UInt32) throws -> UInt32 {
        let result = try throwing { usearch_get(nativeIndex, key, Int(count), vector, USearchScalar.f32.toNative(), $0) }
        return UInt32(result)
    }

    /**
     * @brief Approximate nearest neighbors search.
     * @param vector Single-precision query vector.
     * @param count Upper limit on the number of matches to retrieve.
     * @param filter Closure called for each key, determining whether to include or
     *               skip key in the results.
     * @param keys Optional output buffer for keys of approximate neighbors.
     * @param distances Optional output buffer for (increasing) distances to approximate neighbors.
     * @return Number of matches exported to `keys` and `distances`.
     */
    public func filteredSearchSingle(vector: UnsafePointer<Float32>, count: UInt32, filter: @escaping USearchFilterFn, keys: UnsafeMutablePointer<USearchKey>?, distances: UnsafeMutablePointer<Float32>?) throws -> UInt32 {
        return try filteredSearchGeneric(
            nativeIndex,
            vector: vector,
            count: count,
            quantization: .f32,
            filter: filter,
            keys: keys,
            distances: distances
        )
    }

    /**
     * @brief Adds a labeled vector to the index.
     * @param vector Double-precision vector.
     */
    public func addDouble(key: USearchKey, vector: UnsafePointer<Float64>) throws {
        try throwing { usearch_add(nativeIndex, key, vector, USearchScalar.f64.toNative(), $0) }
    }

    /**
     * @brief Approximate nearest neighbors search.
     * @param vector Double-precision query vector.
     * @param count Upper limit on the number of matches to retrieve.
     * @param keys Optional output buffer for keys of approximate neighbors.
     * @param distances Optional output buffer for (increasing) distances to approximate neighbors.
     * @return Number of matches exported to `keys` and `distances`.
     */
    public func searchDouble(vector: UnsafePointer<Float64>, count: UInt32, keys: UnsafeMutablePointer<USearchKey>?, distances: UnsafeMutablePointer<Float32>?) throws -> UInt32 {
        let found = try throwing { usearch_search(nativeIndex, vector, USearchScalar.f64.toNative(), Int(count), keys, distances, $0) }
        return UInt32(found)
    }

    /**
    * @brief Retrieves a labeled double-precision vector from the index.
    * @param vector A buffer to store the vector.
    * @param count For multi-indexes, the number of vectors to retrieve.
    * @return Number of vectors exported to `vector`.
    */
    public func getDouble(key: USearchKey, vector: UnsafeMutablePointer<Float64>, count: UInt32) throws -> UInt32 {
        let result = try throwing { usearch_get(nativeIndex, key, Int(count), vector, USearchScalar.f64.toNative(),  $0) }
        return UInt32(result)
    }

    /**
     * @brief Approximate nearest neighbors search.
     * @param vector Double-precision query vector.
     * @param count Upper limit on the number of matches to retrieve.
     * @param filter Closure called for each key, determining whether to include or
     *               skip key in the results.
     * @param keys Optional output buffer for keys of approximate neighbors.
     * @param distances Optional output buffer for (increasing) distances to approximate neighbors.
     * @return Number of matches exported to `keys` and `distances`.
     */
    public func filteredSearchDouble(vector: UnsafePointer<Float64>, count: UInt32, filter: @escaping USearchFilterFn, keys: UnsafeMutablePointer<USearchKey>?, distances: UnsafeMutablePointer<Float32>?) throws -> UInt32 {
        return try filteredSearchGeneric(
            nativeIndex,
            vector: vector,
            count: count,
            quantization: .f64,
            filter: filter,
            keys: keys,
            distances: distances
        )
    }

    /**
     * @brief Adds a labeled vector to the index.
     * @param vector Half-precision vector.
     */
    public func addHalf(key: USearchKey, vector: UnsafePointer<Float16>) throws {
        try throwing { usearch_add(nativeIndex, key, vector, USearchScalar.f16.toNative(), $0) }
    }

    /**
     * @brief Approximate nearest neighbors search.
     * @param vector Half-precision query vector.
     * @param count Upper limit on the number of matches to retrieve.
     * @param keys Optional output buffer for keys of approximate neighbors.
     * @param distances Optional output buffer for (increasing) distances to approximate neighbors.
     * @return Number of matches exported to `keys` and `distances`.
     */
    public func searchHalf(vector: UnsafePointer<Float16>, count: UInt32, keys: UnsafeMutablePointer<USearchKey>?, distances: UnsafeMutablePointer<Float32>?) throws -> UInt32 {
        let found = try throwing { usearch_search(nativeIndex, vector, USearchScalar.f16.toNative(), Int(count), keys, distances, $0) }
        return UInt32(found)
    }

    /**
    * @brief Retrieves a labeled half-precision vector from the index.
    * @param vector A buffer to store the vector.
    * @param count For multi-indexes, the number of vectors to retrieve.
    * @return Number of vectors exported to `vector`.
    */
    public func getHalf(key: USearchKey, vector: UnsafeMutablePointer<Float16>, count: UInt32) throws -> UInt32 {
        let result = try throwing { usearch_get(nativeIndex, key, Int(count), vector, USearchScalar.f16.toNative(), $0) }
        return UInt32(result)
    }

    /**
     * @brief Approximate nearest neighbors search.
     * @param vector Double-precision query vector.
     * @param count Upper limit on the number of matches to retrieve.
     * @param filter Closure called for each key, determining whether to include or
     *               skip key in the results.
     * @param keys Optional output buffer for keys of approximate neighbors.
     * @param distances Optional output buffer for (increasing) distances to approximate neighbors.
     * @return Number of matches exported to `keys` and `distances`.
     */
    public func filteredSearchHalf(vector: UnsafePointer<Float16>, count: UInt32, filter: @escaping USearchFilterFn, keys: UnsafeMutablePointer<USearchKey>?, distances: UnsafeMutablePointer<Float32>?) throws -> UInt32 {
        return try filteredSearchGeneric(
            nativeIndex,
            vector: vector,
            count: count,
            quantization: .f16,
            filter: filter,
            keys: keys,
            distances: distances
        )
    }

    public func contains(key: USearchKey) throws -> Bool {
        return try throwing { usearch_contains(nativeIndex, key, $0) }
    }

    public func count(key: USearchKey) throws -> UInt32 {
        return UInt32(try throwing { usearch_count(nativeIndex, key, $0) })
    }

    public func remove(key: USearchKey) throws -> UInt32 {
        return try UInt32(throwing { usearch_remove(nativeIndex, key, $0) })
    }

    public func rename(from key: USearchKey, to newKey: USearchKey) throws -> UInt32 {
        return try UInt32(throwing { usearch_rename(nativeIndex, key, newKey, $0) })
    }

    /**
     * @brief Saves pre-constructed index to disk.
     */
    public func save(path: String) throws {
        guard let cPath = path.cString(using: .utf8) else {
            throw USearchError.pathNotUTF8Encodable
        }
        try throwing { usearch_save(nativeIndex, cPath, $0) }
    }

    /**
     * @brief Loads a pre-constructed index from index.
     */
    public func load(path: String) throws {
        guard let cPath = path.cString(using: .utf8) else {
            throw USearchError.pathNotUTF8Encodable
        }
        try throwing { usearch_load(nativeIndex, cPath, $0) }
    }

    /**
     * @brief Views a pre-constructed index from disk without loading it into RAM.
     *        Allows working with larger-than memory indexes and saving scarce
     *        memory on device in read-only workloads.
     */
    public func view(path: String) throws {
        guard let cPath = path.cString(using: .utf8) else {
            throw USearchError.pathNotUTF8Encodable
        }
        try throwing { usearch_view(nativeIndex, cPath, $0) }
    }

    /**
     * @brief Removes all the data from index, while preserving the settings.
     */
    public func clear() throws {
        try throwing { usearch_clear(nativeIndex, $0) }
    }
}
